### PR TITLE
Make Tests Pass

### PR DESCRIPTION
**Pull request by dlecocq**
Also, add Vagrant. And Travis.

### CODE REVIEW

File: src/test.cpp [hunk 1]
```diff
@@ -120,64 +120,64 @@ TEST_CASE("tokenizers", "We can get tokens from a string") {
...
-        a = hasher(pope.c_str(), pope.length());
-        a = a ^ b;
-        count = 0;
-        while (a) {
-            a = a & (a - 1); ++count;
-        }
-        REQUIRE(count > 5);
-    }
-}
+// TEST_CASE("simhash", "We can perform simhash") {
```
Comment (b4hand) [inline src/test.cpp]: What happened to this test case? Are you going to replace it with anything?

File: src/hash.hpp [hunk 2]
```diff
@@ -13,6 +13,8 @@
 #include "common.h"
 #include "cyclic.hpp"
 
+#include <cstring>
```
Comment (b4hand) [inline src/hash.hpp]: Was this include actually necessary? I don't see you depending on any `str*` functions.

File: src/hash.hpp [hunk 3]
```diff
@@ -24,6 +26,40 @@ namespace Simhash {
...
+    private:
+        class Accumulator {
+        private:
+            int64_t v[64];
+
+            size_t window;
+            Cyclic<hash_t> cyclic;
+
+        public:
+            Accumulator(): window(3), cyclic(window) {
```
Comment (b4hand) [inline src/hash.hpp]: It's fairly typical to include a space before the colon for the initializer list.

File: src/hash.hpp [hunk 4]
```diff
@@ -24,6 +26,40 @@ namespace Simhash {
...
+                memset(v, 0, 64);
+            }
+
+            void update(hash_t hash) {
+                int j(0);
+                for (j = 63; j > 0; --j) {
```
Comment (b4hand) [inline src/hash.hpp]: Declare `j` inside the for `loop`.

File: src/hash.hpp [hunk 5]
```diff
@@ -24,6 +26,40 @@ namespace Simhash {
...
+                    v[j] += (hash & 1) ? 1 : -1;
+                    hash = hash >> 1;
```
Comment (b4hand) [inline src/hash.hpp]:
```c
hash >>= 1;
```

File: src/hash.hpp [hunk 6]
```diff
@@ -24,6 +26,40 @@ namespace Simhash {
...
+                }
+                v[j] += (hash & 1) ? 1 : -1;
```
Comment (b4hand) [inline src/hash.hpp]: Make this a literal `v[0]` once you move the declaration of `j`.

File: src/hash.hpp [hunk 7]
```diff
@@ -24,6 +26,40 @@ namespace Simhash {
...
+            }
+
+            /* With counts appropriately tallied, create a 1 bit for each of
+             * the counts that's positive. That result is the hash. */
+            hash_t result() {
+                hash_t hash(0);
+                for (int j = 0; j < 64; ++j) {
+                    if (v[j] > 0) {
+                        hash = hash | (static_cast<hash_t>(1) << j);
```
Comment (b4hand) [inline src/hash.hpp]:
```c++
hash |= static_cast<hash_t>(1) << j;
```

File: src/hash.hpp (hunk 4 above)
Comment (b4hand) [inline src/hash.hpp]: Also, this loop being `j > 0` is a bit deceptive. Normally it should be `j >= 0` if you count down. I might actually be tempted to make it `j >= 1` just to make it more confirm what it is doing. That's also why it was fine that that was actually an unsigned loop variable before, but it is still strange.

File: src/hash.hpp (hunk 2 above)
Comment (dlecocq) [inline src/hash.hpp]: Yeah, I was using `memset`, but we decided to go `std::vector`.

File: src/hash.hpp (hunk 3 above)
Comment (dlecocq) [inline src/hash.hpp]: Works for me

File: src/hash.hpp (hunk 4 above)
Comment (dlecocq) [inline src/hash.hpp]: Refactored. No reason for it to be as it is now. The reason it was `j > 0` in its previous incarnation is that it was using an unsigned variable.

File: src/hash.hpp (hunk 5 above)
Comment (dlecocq) [inline src/hash.hpp]: Sweet

File: src/hash.hpp (hunk 6 above)
Comment (dlecocq) [inline src/hash.hpp]: Moved inside the loop.

File: src/hash.hpp (hunk 7 above)
Comment (dlecocq) [inline src/hash.hpp]: D'oh

File: src/test.cpp (hunk 1 above)
Comment (dlecocq) [inline src/test.cpp]: Completely forgot to re-enable them. Working on it.

**Comment by dlecocq** [comment]
Reupped.

File: src/hash.hpp [hunk 8]
```diff
@@ -37,94 +69,28 @@ namespace Simhash {
...
+            Accumulator accumulator;
 
-            for (tp = tokens; *tp != NULL; tp++) {
-                hash_t r = cyclic.push(hasher(*tp, strlen(*tp), 0));
-                for (j = 63; j > 0; --j) {
-                    v[j] += (r & 1) ? 1 : -1;
-                    r = r >> 1;
-                }
-                v[j] += (r & 1) ? 1 : -1;
+            for (char **tp = tokens; *tp != NULL; tp++) {
```
Comment (b4hand) [inline src/hash.hpp]: Prefer preincrement over postincrement.

File: src/hash.hpp [hunk 9]
```diff
@@ -37,94 +69,28 @@ namespace Simhash {
...
+            Accumulator accumulator;
 
-            for (i = 0; i < len; i++) {
-                hash_t r = cyclic.push(hasher(reinterpret_cast<char*>(vec+i), sizeof(uint64_t), 0));
-                for (j = 63; j > 0; --j) {
-                    v[j] += (r & 1) ? 1 : -1;
-                    r = r >> 1;
-                }
-                v[j] += (r & 1) ? 1 : -1;
+            for (int i = 0; i < len; i++) {
```
Comment (b4hand) [inline src/hash.hpp]: Same here.

**Comment by b4hand** [comment]
Last minor nits above, and LGTM.

**Comment by b4hand** [comment]
I ended up fixing the preincrement myself in #10.

**Pull request merged**